Assets/UI/worldinput_CQUI.lua

```diff
@@ -0,0 +1,559 @@
+        CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_PASTURE"].Hash);
+        CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_PLANTATION"].Hash);
+        cquiHandledKey = true;
+      end
+    end
+  end
+
+  if (uiKey == Keys.R) then
+    if (unitType == "UNIT_MILITARY_ENGINEER") then
+      -- Build Road/Rail is a UnitOperation
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.UnitOperations["UNITOPERATION_BUILD_ROUTE"].Hash);
+      cquiHandledKey = true;
+    elseif (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.REBASE);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.S and CQUI_isAltDown == true) then
+    if (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestCommand(UI.GetHeadSelectedUnit(), UnitOperationTypes.AIR_ATTACK);
+      -- Use false as we need to let the unmodifed script deal with the Alt Key
+      cquiHandledKey =false;
+    end
+  end
+
+  if uiKey == Keys.VK_SHIFT then
+    CQUI_isShiftDown = false;
+  end
+
+  if uiKey == Keys.VK_ALT then
+    CQUI_isAltDown = false;
+  end
+
+  if (cquiHandledKey == false) then
+    print_debug("CQUI Did not handle key: "..tostring(uiKey));
+    cquiHandledKey = BASE_CQUI_DefaultKeyUpHandler( uiKey );
+  end
+
+  return cquiHandledKey;
+end
+
+-- ===========================================================================
+function OnDefaultKeyDown( pInputStruct:table )
+  print_debug("** Function Entry: OnDefaultKeyDown (CQUI Hook)");
+  if m_isInputBlocked then
+    print_debug("** m_isInputBlocked is true");
+    return;
+  end
+
+  return BASE_CQUI_OnDefaultKeyDown( pInputStruct );
+end
+
+-- ===========================================================================
+function OnDefaultKeyUp( pInputStruct:table )
+  print_debug("** Function Entry: OnDefaultKeyUp (CQUI Hook)");
+  if m_isInputBlocked then
+    print_debug("** m_isInputBlocked is true");
+    return;
+  end
+
+  return BASE_CQUI_OnDefaultKeyUp( pInputStruct );
+end
+
+-- ===========================================================================
+function OnPlacementKeyUp( pInputStruct:table )
+  print_debug("** Function Entry: OnPlacementKeyUp (CQUI Hook)");
+  if m_isInputBlocked then 
+    print_debug("** m_isInputBlocked is true");
+    return;
+  end
+
+  return BASE_CQUI_OnPlacementKeyUp( pInputStruct );
+end
+
+-- ===========================================================================
+function ClearAllCachedInputState()
+  print_debug("** Function Entry: ClearAllCachedInputState (CQUI Hook)");
+  BASE_CQUI_ClearAllCachedInputState();
+
+  CQUI_isShiftDown = false;
+end
+
+-- ===========================================================================
+function OnInputActionTriggered( actionId:number )
+ -- CQUI Enhanced Keyboard over-rides some of the ingame hot keys
+
+end
+
```

Comment:
Unmodded has a new function around here. We should also probably clear our alt key state flag in the same situation
```
-- Alt-F4 in some situations can cause m_isALTDown to be stuck on after you click "Cancel", so clear it here.
function OnUserRequestClose()
  m_isALTDown  = false;
end
```

Reply:
good call